### PR TITLE
Naive sharding

### PR DESCRIPTION
**Pull request by lucksus**
## PR summary

- [x] Push aspects only to nodes that should hold them according to sharding algorithm
- [x] Calculate missing aspects for gossip based on sharding algorithm
- [x] Have `Sime2hWorker` send queries to switch board when full-sync is deactivated
- [x] Have switch board route queries to nodes in the right neighbourhood/shard
- [x] Make tests pass again
- @neonphog - added this to tree as a follow-on item - Have sim2 and sim2h worker negotiate version and if sharding is activated or not 

## testing/benchmarking notes

( if any manual testing or benchmarking was/should be done, add notes and/or screenshots here )

## followups

( any new tickets/concerns that were discovered or created during this work but aren't in scope for review here )

## changelog

- [ ] if this is a code change that effects some consumer (e.g. zome developers) of holochain core,  then it has been added to [our between-release changelog](https://github.com/holochain/holochain-rust/blob/develop/CHANGELOG-UNRELEASED.md) with the format 

```markdown
- summary of change [PR#1234](https://github.com/holochain/holochain-rust/pull/1234)
```

## documentation

- [ ] this code has been documented according to our [docs checklist](https://hackmd.io/@freesig/Hk9AmKJNS)

### CODE REVIEW

File: crates/sim2h/src/cache.rs [hunk 1]
```diff
@@ -166,10 +182,43 @@ impl Space {
...
+        &self,
+        entry_hash: EntryHash,
+        redundant_count: u64,
+    ) -> HashMap<AgentId, AgentInfo> {
+        self.agents
+            .iter()
+            .filter(|(agent, _)| {
+                let agent_id_string: String = (*agent).clone().into();
+                naive_sharding_should_store(
+                    anything_to_location(&self.crypto, &agent_id_string),
```
Comment (neonphog) [inline crates/sim2h/src/cache.rs]: the hashing is not insignificant - we should be caching the location with our agent record

File: crates/sim2h/src/cache.rs [hunk 2]
```diff
@@ -166,10 +182,43 @@ impl Space {
...
     pub fn all_aspects(&self) -> &AspectList {
         &self.all_aspects_hashes
     }
 
+    pub fn aspects_in_shard_for_agent(&self, agent: &AgentId, redundant_count: u64) -> AspectList {
+        self.all_aspects_hashes
+            .filtered_by_entry_hash(|entry_hash| {
+                let agent_id_string: String = agent.clone().into();
+                naive_sharding_should_store(
+                    anything_to_location(&self.crypto, &agent_id_string),
```
Comment (neonphog) [inline crates/sim2h/src/cache.rs]: ditto cache location if possible

File: crates/net/src/sim2h_worker.rs [hunk 3]
```diff
@@ -94,6 +95,7 @@ impl Sim2hWorker {
             initial_authoring_list: None,
             initial_gossiping_list: None,
             has_self_stored_authored_aspects: false,
+            is_full_sync_DHT: true,
```
Comment (neonphog) [inline crates/net/src/sim2h_worker.rs]: needs to match the mode of the sim2h_server we've connected to... we could add a DhtAlgorithm to the same wire-message that lets us compare versions for sim2h compatibility.

File: crates/sim2h/src/cache.rs (hunk 1 above)
Comment (neonphog) [reply crates/sim2h/src/cache.rs]: added to tree

File: crates/sim2h/src/cache.rs (hunk 2 above)
Comment (neonphog) [reply crates/sim2h/src/cache.rs]: added to tree

File: crates/net/src/sim2h_worker.rs (hunk 3 above)
Comment (neonphog) [reply crates/net/src/sim2h_worker.rs]: added to tree

**Comment by lucksus** [comment]
The Sim2h server currently is hard-coded to do sharding: https://github.com/holochain/holochain-rust/pull/1967/commits/6dbd147f59a4b473ac7f69016841fed5f9604194

To properly test this we need the worker in sharded mode too.
Interesting to see though that tests pass when the worker is configured to do full-sync.

 When I looked at sim2h server logs during app spec, my guess was that the query responses somehow dont make their through to core. maybe something in the worker. I could see sim2 handling rhe queries. and since tests pass with sim2h sharded but worker in full-sync, this is pretty much the only difference..

**Comment by zippy** [comment]
as of 509fd9d79fd7 there's a sharding test that at least passes locally.  But, it completely clobbers the sim2h server, which keeps spinning in an odd way after the test completes.

**Comment by zippy** [comment]
b14a78553  adds getting the sharding status from the sim2h_server at worker startup time and setting fullsync or not from that.

**Comment by zippy** [comment]
There is definitely a problem with the sharding implementation.  From the conductor log:

```
ERROR 2019-12-23 11:09:06 [holochain_net::sim2h_worker] net_worker_thread/puid-8-30 crates/net/src/sim2h_worker.rs:414 Got a Lib3hToClientResponse from the Sim2h server, weird! Ignoring: HandleQueryEntryResult(QueryEntryResultData { space_address: SpaceHash(HashString("QmdNj9UTYPGrgUfJWDfF4gWiviJVoDCFQHVWVP9jvQLrXm")), entry_address: EntryHash(HashString("HcSCJ6P5F7UK7BGngzun9G6mb4h3EDo3qMBwx335SuPgif7qN4eyoeX9Eo9rhsa")), request_id: "puid-70-4", requester_agent_id: AgentPubKey(HashString("HcScjCNacf8pyr5pwetyzPpsCiq8ruinfxzIfK6RC87YuiqyS3HXrRvgp7vybyz")), responder_agent_id: AgentPubKey(HashString("HcScI44DNpWO3ArixxCxRy694pkTDuojbMHbDoFrGws7s5rp6jZDR7fbEewnewr")), query_result: "{\"Entry\":{\"entry_with_meta\":{\"entry\":{\"AgentId\":{\"nick\":\"0::3::e0b02fd8-6873-41b4-bcbf-eff5fcb565ec\",\"pub_sign_key\":\"HcSCJ6P5F7UK7BGngzun9G6mb4h3EDo3qMBwx335SuPgif7qN4eyoeX9Eo9rhsa\"}},\"crud_status\":\"live\",\"maybe_link_update_delete\":null},\"headers\":[{\"entry_type\":\"AgentId\",\"entry_address\":\"HcSCJ6P5F7UK7BGngzun9G6mb4h3EDo3qMBwx335SuPgif7qN4eyoeX9Eo9rhsa\",\"provenances\":[[\"HcSCJ6P5F7UK7BGngzun9G6mb4h3EDo3qMBwx335SuPgif7qN4eyoeX9Eo9rhsa\",\"jclGVMKOWHZ8QIRtDQ97LLSEShjy/arv5XWE/iFv3g7kTaKPd+1s5fE855rXeu/tbY1gR0qc1NUifKW0dnSIAQ==\"]],\"link\":\"QmZgQ8tW8ygD5L2dQrBrzykwsC5coKFJCd5bNutiR7bwj4\",\"link_same_type\":null,\"link_update_delete\":null,\"timestamp\":\"2019-12-23T16:07:50+00:00\"}]}}" })
```
Why is sim2h sending the conductor a `Lib3hToClientResponse` when that's what the conductor should be responding with?

File: crates/sim2h/src/naive_sharding.rs [hunk 4]
```diff
@@ -25,20 +24,27 @@ pub fn anything_to_location(crypto: &Box<dyn CryptoSystem>, anything: &str) -> L
...
+
 /// implement a super simple sharding algorithm
 /// to distribute data when node counts go > 50
 pub fn naive_sharding_should_store(
     agent_loc: Location,
     data_addr_loc: Location,
     node_count: u64,
+    redundant_count: u64,
 ) -> bool {
     // if there are < 50 nodes, everyone should store everything
```
Comment (struktured) [inline crates/sim2h/src/naive_sharding.rs]:
```suggestion
    // if there are < `REDUNDANT_COUNT` nodes, everyone should store everything
```

File: crates/sim2h/src/naive_sharding.rs [hunk 5]
```diff
@@ -25,20 +24,27 @@ pub fn anything_to_location(crypto: &Box<dyn CryptoSystem>, anything: &str) -> L
...
-    if node_count <= REDUNDANT_COUNT {
+    if node_count <= redundant_count {
         return true;
     }
 
     // divide up the space so on average data will be stored by 50 nodes
```
Comment (struktured) [inline crates/sim2h/src/naive_sharding.rs]:
```suggestion
    // divide up the space so on average data will be stored by `REDUNDANT_COUNT` nodes
```

**Pull request merged**